### SQL

In [1]:
# Импортируем необходимые библиотеки
import pandas as pd
from sqlalchemy import create_engine
import psycopg2

In [2]:
# Подключаемся к базе
db_config = {'user': '****', # имя пользователя
             'pwd': '****', # пароль
             'host': '****',
             'port': ****, # порт подключения
             'db': '****'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}, echo=False)

**Таблица `visits`**

In [3]:
query = '''
        select * from visits;

        '''

In [4]:
df_visits = pd.io.sql.read_sql(query, con=engine)

In [5]:
df_visits.head()

id                   uid   device               endts  sourceid  \
0   0  16879256277535980062    touch 2017-12-20 17:38:00         4   
1   1    104060357244891740  desktop 2018-02-19 17:21:00         2   
2   2   7459035603376831527    touch 2017-07-01 01:54:00         5   
3   3  16174680259334210214  desktop 2018-05-20 11:23:00         9   
4   4   9969694820036681168  desktop 2017-12-27 14:06:00         3   

              startts  
0 2017-12-20 17:20:00  
1 2018-02-19 16:53:00  
2 2017-07-01 01:54:00  
3 2018-05-20 10:59:00  
4 2017-12-27 14:06:00

In [6]:
df_visits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358532 entries, 0 to 358531
Data columns (total 6 columns):
id          358532 non-null int64
uid         358532 non-null object
device      358532 non-null object
endts       358532 non-null datetime64[ns]
sourceid    358532 non-null int64
startts     358532 non-null datetime64[ns]
dtypes: datetime64[ns](2), int64(2), object(2)
memory usage: 16.4+ MB


**Таблица `orders`**

In [7]:
query = '''
        select * from orders;

        '''
df_orders = pd.io.sql.read_sql(query, con=engine)
df_orders.head()

id               buyts revenue                   uid
0   1 2017-06-01 00:10:00  $17.00  10329302124590727494
1   2 2017-06-01 00:25:00   $0.55  11627257723692907447
2   3 2017-06-01 00:27:00   $0.37  17903680561304213844
3   4 2017-06-01 00:29:00   $0.55  16109239769442553005
4   5 2017-06-01 07:58:00   $0.37  14200605875248379450

In [8]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50415 entries, 0 to 50414
Data columns (total 4 columns):
id         50415 non-null int64
buyts      50415 non-null datetime64[ns]
revenue    50415 non-null object
uid        50415 non-null object
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 1.5+ MB


**Таблица `advertisment_costs`**

In [9]:
query = '''
        select * from advertisment_costs;

        '''
df_adv_cost = pd.io.sql.read_sql(query, con=engine)
df_adv_cost.head()

id  sourceid         dt   costs
0   1         1 2017-06-01  $75.20
1   2         1 2017-06-02  $62.25
2   3         1 2017-06-03  $36.53
3   4         1 2017-06-04  $55.00
4   5         1 2017-06-05  $57.08

In [10]:
df_adv_cost.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2542 entries, 0 to 2541
Data columns (total 4 columns):
id          2542 non-null int64
sourceid    2542 non-null int64
dt          2542 non-null datetime64[ns]
costs       2542 non-null object
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 79.5+ KB


**1. Посчитайте затраты на каждый рекламный источник.**

>Таблица **advertisment_costs** (информация о затратах на маркетинг):
* `sourceid` — идентификатор рекламного источника
* `dt` — дата
* `costs` — затраты на этот рекламный источник в этот день

In [11]:
query_1 = '''
        select
            sourceid, sum(costs) as costs_sum
        from 
            advertisment_costs as adv
        group by adv.sourceid;
        '''

In [12]:
costs_total = pd.io.sql.read_sql(query_1, con=engine)

In [13]:
costs_total

sourceid    costs_sum
0        10    $5,822.49
1         9    $5,517.49
2         1   $20,833.27
3         5   $51,757.10
4         4   $61,073.60
5         2   $42,806.04
6         3  $141,321.63

Больше всего потратили на рекламный источник с номером `3`.

**2. Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы**

In [14]:
query_2 = '''
        SELECT
            MIN(revenue) AS revenue_min,
            MAX(revenue) AS revenue_max,
            SUM(revenue) AS total_revenue      
        FROM
            orders AS o
        WHERE o.uid IN (
                        SELECT DISTINCT uid 
                        FROM visits AS v
                        WHERE v.device='desktop' AND
                              v.sourceid=3
                        );
        '''
df_revenue = pd.io.sql.read_sql(query_2, con = engine)

In [15]:
df_revenue

revenue_min revenue_max total_revenue
0       $0.00   $2,633.28   $106,058.86

>**Выводы**
* Источник рекламы `№ 3` на который потратили больше всего денег `141,321.63 $` принес `106,058.86 $` выручки
* самый "дорогой" заказ принес `2,633.28 $`
* самый "дешевый" не стоил ничего `0.00 $`

**3 Введём новую метрику эффективности рекламны - дневные затраты на рекламу на пользователя по источнику. Она рассчитывается для каждого отдельного дня и равна объёму денежных средств, которые были израсходованы на то, чтобы пользователь перешёл по рекламе на сайт из данного рекламного источника. Иными словами, метрика для каждого дня равна отношению рекламных затрат по данному источнику к количеству посетителей, перешедших на сайт с данного источника. Рассчитайте для каждого дня введённую метрику. Считайте, что целевое действие было произведено тогда, когда сесия началась. Например, если посетитель провёл на сайте несколько минут с 23:50 14 апреля по 0:20 15 апреля, то его нужно включать в рассчёт за 14 апреля**

In [16]:
query_3 = '''
        SELECT 
        adv_cost.dt AS date,
        adv_cost.sourceid,
        adv_cost.daily_cost,
        gr_visits.daily_users,
        adv_cost.daily_cost/gr_visits.daily_users AS ad_costs_per_user
        FROM 
        
        (SELECT
             dt,
             sourceid,
             SUM(costs) AS daily_cost
        FROM 
             advertisment_costs AS adc
        GROUP BY 
             adc.dt,
             adc.sourceid) AS adv_cost
        
        JOIN (SELECT
                    startts, sourceid, COUNT(uid) AS daily_users
              FROM (SELECT DATE_TRUNC('DAY', startts) AS startts, sourceid, uid 
                    FROM visits) AS subquery
              GROUP BY 
                  startts,
                  sourceid               
        ) AS gr_visits
        ON (adv_cost.sourceid = gr_visits.sourceid AND adv_cost.dt = gr_visits.startts)
        ORDER BY date, sourceid;
             
        '''
daily_costs = pd.io.sql.read_sql(query_3, con=engine)
daily_costs

date  sourceid daily_cost  daily_users ad_costs_per_user
0    2017-06-01         1     $75.20           68             $1.10
1    2017-06-01         2    $132.56           85             $1.55
2    2017-06-01         3    $349.38          185             $1.88
3    2017-06-01         4     $76.74          141             $0.54
4    2017-06-01         5     $73.24          121             $0.60
5    2017-06-01         9     $13.06           52             $0.25
6    2017-06-01        10     $15.08            9             $1.67
7    2017-06-02         1     $62.25           60             $1.03
8    2017-06-02         2    $132.38           98             $1.35
9    2017-06-02         3    $314.22          180             $1.74
10   2017-06-02         4     $59.21          158             $0.37
11   2017-06-02         5    $123.82          122             $1.01
12   2017-06-02         9     $18.13           14             $1.29
13   2017-06-02        10     $11.18           21             $0.53
14   2017-06-03         1     $36.53           39             $0.93
15   2017-06-03         2     $77.20           64             $1.20
16   2017-06-03         3    $196.12          124             $1.58
17   2017-06-03         4     $59.39          112             $0.53
18   2017-06-03         5     $64.93           91             $0.71
19   2017-06-03         9      $7.06           22             $0.32
20   2017-06-03        10      $9.62           24             $0.40
21   2017-06-04         1     $55.00           50             $1.10
22   2017-06-04         2     $79.08           59             $1.34
23   2017-06-04         3    $389.87          141             $2.76
24   2017-06-04         4    $161.97          115             $1.40
25   2017-06-04         5     $52.77          110             $0.47
26   2017-06-04         9      $4.65           14             $0.33
27   2017-06-04        10     $17.82           19             $0.93
28   2017-06-05         1     $57.08           85             $0.67
29   2017-06-05         2    $108.96          104             $1.04
...         ...       ...        ...          ...               ...
2511 2018-05-27         9     $11.11            7             $1.58
2512 2018-05-27        10      $9.92           10             $0.99
2513 2018-05-28         1     $54.47          115             $0.47
2514 2018-05-28         2     $82.67          141             $0.58
2515 2018-05-28         3    $247.92          231             $1.07
2516 2018-05-28         4     $68.88          358             $0.19
2517 2018-05-28         5    $131.77          205             $0.64
2518 2018-05-28         9     $12.47           60             $0.20
2519 2018-05-28        10     $21.26           45             $0.47
2520 2018-05-29         1     $61.13          113             $0.54
2521 2018-05-29         2    $105.75          151             $0.70
2522 2018-05-29         3    $347.59          258             $1.34
2523 2018-05-29         4    $103.84          288             $0.36
2524 2018-05-29         5    $138.33          163             $0.84
2525 2018-05-29         9     $16.83           28             $0.60
2526 2018-05-29        10     $11.32           32             $0.35
2527 2018-05-30         1    $130.47          123             $1.06
2528 2018-05-30         2    $119.26          362             $0.32
2529 2018-05-30         3    $611.78          339             $1.80
2530 2018-05-30         4     $18.01          327             $0.05
2531 2018-05-30         5    $246.76          179             $1.37
2532 2018-05-30         9     $23.57            9             $2.61
2533 2018-05-30        10     $33.15           69             $0.48
2534 2018-05-31         1    $141.70          241             $0.58
2535 2018-05-31         2    $254.15          642             $0.39
2536 2018-05-31         3    $822.30          404             $2.03
2537 2018-05-31         4    $492.54          508             $0.96

>**Вывод**
* Стоимость привлечения пользователя с каждого источника меняется каждый день  
* Источник `№ 3`стабильно дорогой
* Более выгодными представляются источники `№ 2` и `№ 4`
